packages/app/src/vis-packs/core/heatmap/MappedHeatmapVis.tsx

```diff
@@ -58,22 +64,47 @@ function MappedHeatmapVis(props: Props) {
     customDomain,
     colorMap,
     scaleType,
+    complexVisType,
     keepRatio,
     showGrid,
     invertColorMap,
     flipXAxis,
     flipYAxis,
   } = config;
 
-  const numArray = useToNumArray(value);
+  const { shape: dims, type } = dataset;
+  const [slicedDims, slicedMapping] = useSlicedDimsAndMapping(dims, dimMapping);
+  const isComplex = isComplexArray(value);
+
+  const numArray = useMemo(
+    () => (!isComplex ? toNumArray(value) : undefined),
+    [isComplex, value],
+  );
+  const [phaseArray, amplitudeArray] = useMemo(
+    () => (isComplex ? getPhaseAmp(value) : []),
+    [isComplex, value],
+  );
+
   const numAxisArrays = useToNumArrays(axisValues);
 
-  const { shape: dims } = dataset;
-  const [slicedDims, slicedMapping] = useSlicedDimsAndMapping(dims, dimMapping);
-  const dataArray = useMappedArray(numArray, slicedDims, slicedMapping);
+  const hookArgs = [slicedDims, slicedMapping] as const;
+  const numMapped = useMappedArray(numArray, ...hookArgs);
+  const phaseMapped = useMappedArray(phaseArray, ...hookArgs);
+  const amplitudeMapped = useMappedArray(amplitudeArray, ...hookArgs);
+
+  const numDomain =
+    useDomain(numMapped, scaleType, undefined, ignoreValue) || DEFAULT_DOMAIN;
+  const phaseDomain = useDomain(phaseMapped, scaleType) || DEFAULT_DOMAIN;
+  const amplitudeDomain =
+    useDomain(amplitudeMapped, scaleType) || DEFAULT_DOMAIN;
+
+  const [dataArray, dataDomain] = numMapped
+    ? [numMapped, numDomain]
+    : complexVisType === ComplexVisType.Amplitude
+      ? [amplitudeMapped, amplitudeDomain]
+      : [phaseMapped, phaseDomain];
 
-  const dataDomain =
-    useDomain(dataArray, scaleType, undefined, ignoreValue) || DEFAULT_DOMAIN;
+  assertDefined(dataArray);
```

Comment:
TypeScript is not clever enough to realise that `dataArray` is always defined here.